### PR TITLE
Add managed identity, dapr enable and application insights to container apps

### PR DESCRIPTION
**Pull request by mukundansundar**
- This PR adds a daprEnabled flag to be set if a container App has dapr integration.
- This PR also add user defined managed Identity name as a parameter for container App bicep script.
- Have marked old parameter `managedIdentity` as deprecated to be removed later (Can be removed now if breaking changes are ok)
- Added Application Insights Dapr link in Container Apps if both are defined

### CODE REVIEW

File: templates/common/infra/bicep/core/host/container-app.bicep [hunk 1]
```diff
@@ -5,13 +5,33 @@ param tags object = {}
 param containerAppsEnvironmentName string
 param containerName string = 'main'
 param containerRegistryName string
+
+@description('Minimum number of replicas to run')
```
Comment (ellismg) [inline templates/common/infra/bicep/core/host/container-app.bicep:9]: Should we use the `@minValue` decorator here to enforce that the value is >= 0?

File: templates/common/infra/bicep/core/host/container-app.bicep [hunk 2]
```diff
@@ -58,6 +90,10 @@ resource app 'Microsoft.App/containerApps@2022-03-01' = {
           }
         }
       ]
+      scale: {
+        minReplicas: containerMaxReplicas > 0 ? containerMaxReplicas : 0
```
Comment (ellismg) [inline templates/common/infra/bicep/core/host/container-app.bicep:94]: I think we might have inverted these?  I would have expected that `minReplicas`'s value would be set based on `containerMinReplicas` not "Max" as you have here.  If we use the `@minValue` decorators on the parameters (as I noted above) to enforce these conditions, I think we can get rid of this logic here.

File: templates/common/infra/bicep/core/host/container-app.bicep [hunk 3]
```diff
@@ -5,13 +5,33 @@ param tags object = {}
...
+@description('Minimum number of replicas to run')
+param containerMinReplicas int = 1
+@description('Maximum number of replicas to run')
+param containerMaxReplicas int = 1
+
 param env array = []
 param external bool = true
 param imageName string
 param keyVaultName string = ''
+@description('Param to denote whether to use a managed identity or not. If not, the identityPrincipalId output will be empty. DEPRECATED: Use managedIdentityEnabled instead.')
```
Comment (ellismg) [inline templates/common/infra/bicep/core/host/container-app.bicep:18]: Perhaps a question for @jongio - should we just remove this deprecated property as part of this PR? If not, what's the plan to remove it long term?

File: templates/common/infra/bicep/core/host/container-app.bicep [hunk 4]
```diff
@@ -5,13 +5,33 @@ param tags object = {}
 param containerAppsEnvironmentName string
 param containerName string = 'main'
 param containerRegistryName string
+
+@description('Minimum number of replicas to run')
+param containerMinReplicas int = 1
+@description('Maximum number of replicas to run')
```
Comment (ellismg) [inline templates/common/infra/bicep/core/host/container-app.bicep:11]: Similar feedback, perhaps we should use `@minValue` here to enforce this is always 1 or more?

File: templates/common/infra/bicep/core/host/container-app.bicep (hunk 2 above)
Comment (mukundansundar) [reply templates/common/infra/bicep/core/host/container-app.bicep:94]: fixed. removed the condition here and added minValue constraints to the params.

File: templates/common/infra/bicep/core/host/container-app.bicep (hunk 4 above)
Comment (mukundansundar) [reply templates/common/infra/bicep/core/host/container-app.bicep:11]: done

File: templates/common/infra/bicep/core/host/container-app.bicep (hunk 1 above)
Comment (mukundansundar) [reply templates/common/infra/bicep/core/host/container-app.bicep:9]: done

File: templates/common/infra/bicep/core/host/container-app.bicep (hunk 3 above)
Comment (jongio) [reply templates/common/infra/bicep/core/host/container-app.bicep:20]: I agree it is a better name.  If we take this, then we'll want to update all modules to use the same name.  This includes all bicep and terraform in azure-dev.  That can come as a diff PR if too large here, but don't just change it here.

File: templates/common/infra/bicep/core/host/container-app.bicep [hunk 5]
```diff
@@ -71,8 +107,15 @@ resource containerRegistry 'Microsoft.ContainerRegistry/registries@2022-02-01-pr
   name: containerRegistryName
 }
 
+// user assigned managed identity to use throughout
+resource managedIdentityRes 'Microsoft.ManagedIdentity/userAssignedIdentities@2023-01-31' existing = {
```
Comment (jongio) [inline templates/common/infra/bicep/core/host/container-app.bicep:113]: Do you need Res in name

File: templates/common/infra/bicep/core/host/container-app.bicep (hunk 3 above)
Comment (jongio) [reply templates/common/infra/bicep/core/host/container-app.bicep:20]: ...and don't have both...

File: templates/common/infra/bicep/core/host/container-app.bicep [hunk 6]
```diff
@@ -71,8 +107,15 @@ resource containerRegistry 'Microsoft.ContainerRegistry/registries@2022-02-01-pr
...
+  name: managedIdentityName
+}
+
+
 output defaultDomain string = containerAppsEnvironment.properties.defaultDomain
 output identityPrincipalId string = managedIdentity ? app.identity.principalId : ''
+output userManagedIdentityId string = managedIdentityEnabled ? managedIdentityRes.id : ''
```
Comment (jongio) [inline templates/common/infra/bicep/core/host/container-app.bicep:120]: I don't think we need this extra output

File: templates/common/infra/bicep/core/host/container-apps-environment.bicep [hunk 7]
```diff
@@ -16,12 +18,17 @@ resource containerAppsEnvironment 'Microsoft.App/managedEnvironments@2022-03-01'
         sharedKey: logAnalyticsWorkspace.listKeys().primarySharedKey
       }
     }
+    daprAIInstrumentationKey: daprEnabled && applicationInsightsName != '' ? applicationInsights.properties.InstrumentationKey : ''
```
Comment (jongio) [inline templates/common/infra/bicep/core/host/container-apps-environment.bicep:21]: Is this a container apps thing to have a special prop for dapr instrumentation?  Seems odd that dapr would be in the name and not use a generic instrumentation key name.

**Review by jongio — CHANGES_REQUESTED** [review]

File: templates/common/infra/bicep/core/host/container-apps-environment.bicep (hunk 7 above)
Comment (ahmelsayed) [reply templates/common/infra/bicep/core/host/container-apps-environment.bicep:21]: yes, there is indeed a special prop for dapr instrumentation on the managedEnvironment resource.

File: templates/common/infra/bicep/core/host/container-app.bicep (hunk 5 above)
Comment (mukundansundar) [reply templates/common/infra/bicep/core/host/container-app.bicep:113]: I needed to differentiate between `managedIdentityRes`(Resource) and the `managedIdentity` deprecated flag. If `managedIdentity` was not a flag would have used that name instead. Do you have any recommendations for the name?

File: templates/common/infra/bicep/core/host/container-app.bicep (hunk 6 above)
Comment (mukundansundar) [reply templates/common/infra/bicep/core/host/container-app.bicep:120]: will remove.

File: templates/common/infra/bicep/core/host/container-app.bicep [hunk 8]
```diff
@@ -5,13 +5,35 @@ param tags object = {}
...
+param managedIdentityEnabled bool = !empty(keyVaultName)
+@description('Name of the managed identity to use.')
+param managedIdentityName string = ''
+
 param targetPort int = 80
 
+@description('Enable Dapr')
+param daprEnabled bool = false
+@description('Dapr app ID')
+param daprApp string = containerName
```
Comment (ahmelsayed) [inline templates/common/infra/bicep/core/host/container-app.bicep:32]:
```suggestion
param daprAppId string = containerName
```

File: templates/common/infra/bicep/core/host/container-app.bicep [hunk 9]
```diff
@@ -22,7 +44,13 @@ resource app 'Microsoft.App/containerApps@2022-03-01' = {
   name: name
   location: location
   tags: tags
-  identity: { type: managedIdentity ? 'SystemAssigned' : 'None' }
+  identity: managedIdentityEnabled || managedIdentity ? {
+    type: 'SystemAssigned,UserAssigned'
```
Comment (ahmelsayed) [inline templates/common/infra/bicep/core/host/container-app.bicep:48]: is it ok to always have `UserAssigned` here even if it's only system assigned identity?

File: templates/common/infra/bicep/core/host/container-app.bicep [hunk 10]
```diff
@@ -71,8 +109,15 @@ resource containerRegistry 'Microsoft.ContainerRegistry/registries@2022-02-01-pr
   name: containerRegistryName
 }
 
+// user assigned managed identity to use throughout
+resource managedIdentityRes 'Microsoft.ManagedIdentity/userAssignedIdentities@2023-01-31' existing = {
+  name: managedIdentityName
```
[the remark this answers is not in the document]
Comment (mukundansundar) [reply templates/common/infra/bicep/core/host/container-app.bicep:114]: Have added a conditional check

File: templates/common/infra/bicep/core/host/container-app.bicep (hunk 5 above)
Comment (mukundansundar) [reply templates/common/infra/bicep/core/host/container-app.bicep:113]: using `userManagedIdentity` for now.

File: templates/common/infra/bicep/core/host/container-app.bicep (hunk 8 above)
Comment (mukundansundar) [reply templates/common/infra/bicep/core/host/container-app.bicep:32]: changed

File: templates/common/infra/bicep/core/host/container-app.bicep (hunk 9 above)
Comment (mukundansundar) [reply templates/common/infra/bicep/core/host/container-app.bicep:48]: For this case, I can add a conditional ternary operator again checking on the managedIdentity name, if provided assign both `UserAssigned` else if managed identity alone is enabled use `SystemAssigned` else have it as none. 
There is a behavior that I noticed that if both `SystemAssigned,UserAssigned` is used, `SystemAssigned` take more priority and that might be an issue.

**Review by ahmelsayed — APPROVED** [review]

File: templates/common/infra/bicep/core/host/container-app.bicep (hunk 3 above)
Comment (jongio) [reply templates/common/infra/bicep/core/host/container-app.bicep:20]: No, let's not keep it.  It's a breaking change that we'll need to think about and flesh out.

**Comment by wbreza** [comment]
Replaced by #2157

**Pull request closed without merging**